service/internal/graph/graph_test.go

```diff
@@ -651,6 +651,30 @@ func TestConnectorPipelinesGraph(t *testing.T) {
 				}
 				assert.Equal(t, expectMutatesData, pipeline.capabilitiesNode.getConsumer().Capabilities().MutatesData)
 
+				// A connector that acts as a receiver in a pipeline that mutates data must inherit mutability
+				if expectMutatesData {
+					for _, receiver := range pipeline.receivers {
+						r, ok := receiver.(*connectorNode)
+						if !ok {
+							continue
+						}
+						for expPipelineID, expPipeline := range pg.pipelines {
+							if expPipelineID.Type() != pipelineID.Type() {
+								continue
+							}
+							for _, exporter := range expPipeline.exporters {
+								e, ok := receiver.(*connectorNode)
+								if !ok || exporter.ID() != receiver.ID() {
+									continue
+								}
+								assert.True(t, e.getConsumer().Capabilities().MutatesData,
```

Comment:
Could this be rewritten as `assert.Equal(t, expectMutatesData, e.getConsumer().Capabilities().MutatesData)` and avoid the conditional at line 655? Thinking this would prevent regressions where a consumer reports mutable when it's not expected

Reply:
The problem is that this assertion is not necessary expected. This test case contains an example of where it goes wrong:

```go
{
	name: "pipelines_conn_fork_merge_traces.yaml",
	pipelineConfigs: map[component.ID]*PipelineConfig{
		component.NewIDWithName("traces", "in"): {
			Receivers:  []component.ID{component.NewID("examplereceiver")},
			Processors: []component.ID{component.NewID("exampleprocessor")},
			Exporters:  []component.ID{component.NewIDWithName("exampleconnector", "fork")}, // wrapped w/ mutates: true
		},
		component.NewIDWithName("traces", "type0"): {
			Receivers:  []component.ID{component.NewIDWithName("exampleconnector", "fork")},
			Processors: []component.ID{component.NewID("exampleprocessor")},
			Exporters:  []component.ID{component.NewIDWithName("exampleconnector", "merge")},
		},
		component.NewIDWithName("traces", "type1"): {
			Receivers:  []component.ID{component.NewIDWithName("exampleconnector", "fork")},
			Processors: []component.ID{component.NewIDWithName("exampleprocessor", "mutate")},
			Exporters:  []component.ID{component.NewIDWithName("exampleconnector", "merge")},
		},
		component.NewIDWithName("traces", "out"): {
			Receivers:  []component.ID{component.NewIDWithName("exampleconnector", "merge")},
			Processors: []component.ID{component.NewID("exampleprocessor")},
			Exporters:  []component.ID{component.NewID("exampleexporter")},
		},
	},
	expectedPerExporter: 2,
},
```

The `type0` and `type1` pipelines both use the `/fork` connector as a receiver, but only `type1` mutates data. So the expectation is that the connector will inherit `MutatesData: true` from `type1`. However, `expectMutatesData` is specific to each pipeline, so when looking at `type0`, we don't expect a match. Essentially, we can't prove a negative (without looking at the broader picture).

All that said, your point is valid and I've reworked this entire section in a way that accounts for all pipelines to which a connector emits, which allows us to validate both ways.